File: src/Cdd.Core/History.fs
namespace Cdd.Core

/// Modell-Historie: weil jeder SPOT-Knoten ein git-diffbares JSON-File unter .spot/ ist,
/// IST `git log` über .spot/ die Historie des Modells — Zeitreise gratis, ohne eigenen Store.
/// Read-only und defensiv: kein git / kein Repo → [] bzw. "" statt Exception (wie /api/infra/status).
module History =

    open System.Diagnostics
    open System.Text.RegularExpressions

    /// Ein Commit, der das Modell (mit-)veränderte.
    type Commit =
        { Hash: string; Short: string; Author: string; Date: string; Subject: string }

    let private idOk = Regex("^[a-zA-Z0-9][a-zA-Z0-9_-]*$")
    let private hashOk = Regex("^[0-9a-fA-F]{4,40}$")

    // Trennzeichen, die in Commit-Nachrichten nicht vorkommen → kein Quoting-Chaos.
    let private US = ""   // zwischen Feldern
    let private RS = ""   // zwischen Commits
    let private fmt = sprintf "--pretty=format:%%H%s%%h%s%%an%s%%ad%s%%s%s" US US US US RS

    /// `git -C <root> <args>`; stdout bei Erfolg, sonst "" (nie throw).
    let private git (root: string) (args: string list) : string =
        try
            let psi = ProcessStartInfo("git")
            psi.ArgumentList.Add("-C"); psi.ArgumentList.Add(root)
            args |> List.iter psi.ArgumentList.Add
            psi.RedirectStandardOutput <- true
            psi.RedirectStandardError <- true
            psi.UseShellExecute <- false
            use p = new Process()
            p.StartInfo <- psi
            if not (p.Start()) then ""
            else
                let out = p.StandardOutput.ReadToEnd()
                p.WaitForExit()
                if p.ExitCode = 0 then out else ""
        with _ -> ""

    let private parse (raw: string) : Commit list =
        if System.String.IsNullOrWhiteSpace raw then []
        else
            raw.Split([| RS |], System.StringSplitOptions.RemoveEmptyEntries)
            |> Array.choose (fun rc ->
                let f = rc.Trim().Split([| US |], System.StringSplitOptions.None)
                if f.Length >= 5 then
                    Some { Hash = f.[0]; Short = f.[1]; Author = f.[2]; Date = f.[3]; Subject = f.[4] }
                else None)
            |> Array.toList

    /// Commits, die irgendeinen .spot/-Knoten berührten — die Modell-Timeline.
    let model (root: string) (limit: int) : Commit list =
        let n = max 1 (min limit 500)
        git root [ "log"; sprintf "-%d" n; "--date=short"; fmt; "--"; ".spot" ] |> parse

    /// Commits, die EINEN Knoten berührten (Knoten-Lebenslauf).
    let node (root: string) (id: string) (limit: int) : Commit list =
        if not (idOk.IsMatch id) then []
        else
            let n = max 1 (min limit 200)
            git root [ "log"; sprintf "-%d" n; "--date=short"; "--follow"; fmt; "--"; ".spot/" + id + ".json" ] |> parse

    /// Der rohe Knoten-JSON-Stand zu einem Commit (Zeitreise). Sanitisiert gegen Path-/Ref-Injection.
    let nodeAt (root: string) (id: string) (hash: string) : string =
        if not (idOk.IsMatch id) || not (hashOk.IsMatch hash) then ""
        else git root [ "show"; hash + ":.spot/" + id + ".json" ]

File: docs/ide/history.js
// Historie: git log über .spot/ — die Modell-Zeitreise (jeder Knoten ein git-diffbares JSON-File).
// Klick auf einen Commit lässt den Faden die Änderung erklären (chat-primär bewahrt).
import { api, escapeHtml } from './core.js';

export function renderHistory(el, store, actions) {
  el.innerHTML = `
    <div class="stage-hint">Modell-Historie — jeder SPOT-Knoten ist ein git-diffbares JSON-File, also IST <code>git log</code> über <code>.spot/</code> die Historie des Modells. Klick einen Commit → der Faden erklärt die Änderung.</div>
    <div class="hist-list" id="hist-list"><div class="muted pad">lade…</div></div>`;
  const list = el.querySelector('#hist-list');
  api.history(80).then(commits => {
    if (!Array.isArray(commits) || !commits.length) {
      list.innerHTML = '<div class="muted pad">Keine git-Historie (kein Repo / kein git).</div>'; return;
    }
    list.innerHTML = commits.map(c =>
      `<div class="hist-row" data-h="${escapeHtml(c.Short)}" data-subj="${escapeHtml(c.Subject)}">
        <span class="hist-date">${escapeHtml(c.Date)}</span>
        <code class="hist-hash">${escapeHtml(c.Short)}</code>
        <span class="hist-subj">${escapeHtml(c.Subject)}</span>
        <span class="hist-author">${escapeHtml(c.Author)}</span></div>`).join('');
    list.querySelectorAll('.hist-row').forEach(r => r.onclick = () =>
      actions.ask(`Was änderte sich im Modell mit Commit ${r.dataset.h} („${r.dataset.subj}")? Fasse die SPOT-Änderung knapp zusammen.`));
  }).catch(() => { list.innerHTML = '<div class="muted pad">Historie nicht erreichbar.</div>'; });
}
